base/compiler/tfuncs.jl

```diff
@@ -1997,6 +1997,17 @@ function tuple_tfunc(𝕃::AbstractLattice, argtypes::Vector{Any})
     return anyinfo ? PartialStruct(typ, argtypes) : typ
 end
 
+@nospecs function memorynew_tfunc(𝕃::AbstractLattice, memtype, m)
+    hasintersect(widenconst(m), Int) || return Bottom
+    if isa(memtype, Const)
+        mem = memtype.val
+        mem <: GenericMemory || return Bottom
+        return mem
+    end
+    return GenericMemory
```

Comment:
(aka, this is the same function as `new` followed by `typeassert`)
```suggestion
    return tmeet(𝕃, instanceof_tfunc(memtype, true)[1], GenericMemory)
```